protocol-specs/generators/bft_processing/index.js

```diff
@@ -313,10 +314,202 @@ const bftInvalidBlockHeaderTestSuiteGenerator = ({
 
 		// invalidPreviouslyForgedHeight(activeDelegates),
 
-		invalidLowermaxHeightPrevoted(activeDelegates),
+		invalidLowerMaxHeightPrevoted(activeDelegates),
 	],
 });
 
+const FORK_STATUS_IDENTICAL_BLOCK = 1;
+const FORK_STATUS_VALID_BLOCK = 2;
+const FORK_STATUS_DOUBLE_FORGING = 3;
+const FORK_STATUS_TIE_BREAK = 4;
+const FORK_STATUS_DIFFERENT_CHAIN = 5;
+const FORK_STATUS_DISCARD = 6;
+
+const bftForkChoiceTestSuiteGenerator = () => {
+	const blockInterval = 10;
+	const lastBlockHeight = 10;
+	const epochTime = Math.floor(new Date('2016-05-24T17:00:00.000Z').getTime());
+
+	// All times are epoch time
+	const lastBlock = generateBlockForBFT({
+		height: 10,
+		timestamp: (lastBlockHeight - 1) * blockInterval, // Block slot time was height * blockInterval
+		receivedAt: (lastBlockHeight - 1) * blockInterval + 2, // Block received 2 seconds after its slot time started
+	});
+
+	const initialState = {
+		blockInterval,
+		lastBlock,
+		epochTime,
+	};
+
+	return {
+		title: 'BFT processing generation',
+		summary: 'Generate set of blocks to verify fork choice rules',
+		config: {
+			forkStatuses: {
+				FORK_STATUS_IDENTICAL_BLOCK,
+				FORK_STATUS_VALID_BLOCK,
+				FORK_STATUS_DOUBLE_FORGING,
+				FORK_STATUS_TIE_BREAK,
+				FORK_STATUS_DIFFERENT_CHAIN,
+				FORK_STATUS_DISCARD,
+			},
+		},
+		runner: 'bft_processing',
+		handler: 'bft_fork_choice_rules',
+		testCases: [
+			{
+				description:
+					'IDENTICAL_BLOCK: Received identical block, as described as "Case 1" in the LIP',
+				initialState,
+				input: {
+					// Block id is the only check to match identical blocks
+					receivedBlock: { ...lastBlock },
+				},
+				output: {
+					forkStatus: FORK_STATUS_IDENTICAL_BLOCK,
+				},
+			},
+			{
+				description:
+					'VALID_BLOCK: Received valid block, as described as "Case 2" in the LIP',
+				initialState,
+				input: {
+					// Valid blocks are always one step ahead and linked to previous block
+					receivedBlock: generateBlockForBFT({
+						height: lastBlock.height + 1,
+						previousBlockId: lastBlock.id,
+					}),
+				},
+				output: {
+					forkStatus: FORK_STATUS_VALID_BLOCK,
+				},
+			},
+			{
+				description:
+					'DISCARD: Received invalid block for current state of chain',
+				initialState,
+				input: {
+					// Any block with lower height than last block is invalid
```

Comment:
this comment comment can be improved because if the `maxHeightPrevoted` is higher but height is lower, then it will be `different chain`